File: 02-week-2--recursion/02-day-2--recursive-search/javascript/recursive_search.js
function recursiveSearch(arr, target) {
  // base cases
  if(arr.length === 0) {
    return false;
  }
  if(arr[0] === target) {
    return true;
  }

  return recursiveSearch(arr.slice(1), target);
}

if (require.main === module) {
  console.log("Expecting: true");
  console.log("=>", recursiveSearch([1, 2, 3], 2));

  console.log("");

  console.log("Expecting: false");
  console.log("=>", recursiveSearch([3, 2, 1], 4));
}

module.exports = recursiveSearch;

// Recursive Search

// Given an Array of values, use recursion to find the target value. 
// Return `true` if found, otherwise `false`.

// ```
// Input: [1, 2, 3], 2
// Output: true

// Input: [3, 2, 1], 4
// Output: false
// ```

// Iterative solution:
// function iterativeSearch(arr, target) {
//   for (const value of arr) {
//     if (value === target) {
//       return true;
//     }
//   }

//   return false;
// }

File: 02-week-2--recursion/02-day-2--recursive-search/ruby/recursive_search.rb
def recursive_search(arr, target)
  # base cases
  return false if arr.empty?
  return true if arr.first == target
  
  recursive_search(arr[1..-1], target)
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: true"
  puts "=>", recursive_search([1, 2, 3], 2)

  puts

  puts "Expecting: false"
  puts "=>", recursive_search([3, 2, 1], 4)
end

# Recursive Search

# Given an Array of values, use recursion to find the target value. 
# Return `true` if found, otherwise `false`.

# ```
# Input: [1, 2, 3], 2
# Output: true

# Input: [3, 2, 1], 4
# Output: false
# ```

# Iterative solution:
# def iterative_search(arr, target)
#   arr.each do |value|
#     return true if value == target
#   end

#   false
# end
